### PR TITLE
use iterators for split

### PR DESCRIPTION
**Pull request by Karsten1987**
I am extending the current split implementation for a iterator based design. This allows to use the same functionality for split to be used in different containers (e.g. `std::list`, `std::set`) and gives the opportunity for custom iterators. Especially the latter seems quite useful in order to correctly split and parse a message definition and extract it as a tuple.

CI: (build: up-to rcpputils, test: packages-select rcpputils)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/7342/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3507/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6020/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7163/)

Signed-off-by: Karsten Knese <karsten@openrobotics.org>

### CODE REVIEW

File: include/rcpputils/split.hpp [hunk 1]
```diff
@@ -36,34 +36,57 @@
...
  * \param[in] input the input string to be split
  * \param[in] delim the dlelimiter used to split the input string
- * \return A vector of tokens.
+ * \param[in] insert iterator pointing to a storage contianer
  */
-inline std::vector<std::string>
-split(const std::string & input, char delim, bool skip_empty = false)
+template<
+  class InsertIterator,
+  typename std::enable_if<
```
Comment (thomas-moulard) [inline include/rcpputils/split.hpp]: Could static_assert be used here instead to improve readability?

File: include/rcpputils/split.hpp [hunk 2]
```diff
@@ -36,34 +36,57 @@
...
+    std::is_same<
+      InsertIterator &,
+      decltype(std::declval<InsertIterator>().operator=(std::declval<std::string>()))>::value
+  >::type * = nullptr>
+inline void
```
Comment (thomas-moulard) [inline include/rcpputils/split.hpp]: inline is not needed anymore here

File: test/test_split.cpp [hunk 3]
```diff
@@ -221,3 +228,73 @@ TEST(test_split, split_backslash) {
     EXPECT_EQ("world", ret[5]);
   }
 }
+
+TEST(test_split, iterator)
```
Comment (thomas-moulard) [inline test/test_split.cpp]: This could be three test cases.

File: test/test_split.cpp [hunk 4]
```diff
@@ -221,3 +228,73 @@ TEST(test_split, split_backslash) {
...
+    return *this;
+  }
+
+  std::tuple<std::string, std::string, std::string> get() const
+  {
+    return std::tie(package_name, middle_element, message_name);
+  }
+};
+
+TEST(test_split, custom_iterator)
```
Comment (thomas-moulard) [inline test/test_split.cpp]: It'd be nice to also have a test case for the error case with EXPECT_THROW.

File: include/rcpputils/split.hpp (hunk 1 above)
Comment (Karsten1987) [reply include/rcpputils/split.hpp]: The reason I opted for SFINAE is that `static_assert` doesn't participate in function overloading. That being, the current template approach has the benefit of having another function declaration set with `std::enable_if` which accepts a different form of iterator. One of them could be just a class member function, e.g. `add` which takes the string and operates on it.
Does this make sense?

**Comment by Karsten1987** [comment]
I hopefully addressed all your comments. I am happy to re-iterate over `static_assert` vs SFINAE. I don't necessarily feel super strong about it.

File: include/rcpputils/split.hpp (hunk 1 above)
Comment (thomas-moulard) [reply include/rcpputils/split.hpp]: I see, that makes sense. I just feel it's hard to read and I wished there was a better way to write this, but I don't see any right now so this LGTM ;)

**Pull request merged**